File: system/cgi/hello.php
#!/usr/bin/php-cgi

<?php

$request_method = getenv('REQUEST_METHOD');
$output_file = getenv('OUTPUT_FILE');

if ($request_method === 'GET')
{
	// get query string from env
	$query_string = getenv('QUERY_STRING');
	
	// parse the query string into key value pairs
	parse_str($query_string, $query_vars);
	
	// extract value for specific keys into vars
	$first_name = $query_vars['first_name'] ?? 'none';
	$last_name = $query_vars['last_name'] ?? 'none';
}

if ($request_method === 'POST')
{
	$input_file = getenv('INPUT_FILE');
	$input_data = file_get_contents($input_file);

	parse_str($input_data, $post_vars);

	$first_name = $post_vars['first_name'] ?? 'none';
	$last_name = $post_vars['last_name'] ?? 'none';
}

$html_content = <<<HTML
<!DOCTYPE html>
<html>

<head>
	<title>PHP CGI Example</title>
	<link rel="stylesheet" type="text/css" href="/styles.css"/>
</head>

<body>
	<h1>PHP CGI:<br>$request_method</h1>
	<p>Hello $first_name $last_name!</p>
</body>

</html>
HTML;

file_put_contents($output_file, $html_content);

?>

File: system/cgi/hello.py
#!/usr/bin/python3

# Import modules for CGI handling 
import cgi, cgitb, os

request_method = os.environ.get("REQUEST_METHOD", "N/A")
output_file = os.environ.get("OUTPUT_FILE", "")
input_file = os.environ.get("INPUT_FILE", "")

if request_method == "GET":
	form = cgi.FieldStorage()
	first_name = form.getvalue('first_name')
	last_name  = form.getvalue('last_name')

elif request_method == "POST":
    with open(input_file, 'r') as file:
        input_data = file.read()
    input_data = input_data.strip()
    input_pairs = input_data.split('&')
    input_dict = {pair.split('=')[0]: pair.split('=')[1] for pair in input_pairs}
    first_name = input_dict.get('first_name', '')
    last_name = input_dict.get('last_name', '')


html_content = """
<!DOCTYPE html>
<html>

<head>
	<title>Python CGI Example</title>
	<link rel="stylesheet" type="text/css" href="/styles.css"/>
</head>

<body>
	<h1>Python CGI:<br>{}</h1>
	<p>Hello {} {}!</p>
</body>

</html>
"""
with open(output_file, 'w') as file:
        file.write(html_content.format(request_method, first_name, last_name))